Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/VideoProcessor.java

```diff
@@ -0,0 +1,119 @@
+  private long recordStartTime = CameraFragment.Companion.getRecordingStartMillis();
+  private long frameDuration = 1000L / CameraFragment.Companion.getFpsRecording();
+  private long framesProcessed = 0;
+
+  public ArrayList<Long> getVideoFrameTimestamp() { return videoFrameTimestamp; }
+
+  /** Sets media-reader and loads available video frames. SAFE_FRAMES trims potential corrupted
+   * frames from the end.*/
+  @RequiresApi(api = VERSION_CODES.P)
+  public boolean createVideoReader(Context applicationContext, Uri fileUri) {
+    recognizer = TextRecognition.getClient();
+    MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever();
+    mediaMetadataRetriever.setDataSource(applicationContext, fileUri);
+    int totalFrames = Integer
+        .parseInt(mediaMetadataRetriever.extractMetadata(METADATA_KEY_VIDEO_FRAME_COUNT));
+
+    // TODO: Add on demand frame read opposed to read all frames.
+    for (int i = 0; i < totalFrames; i += FRAME_CHUNK_READ_SIZE) {
+      try {
+        frameList.addAll(
+            mediaMetadataRetriever.getFramesAtIndex(
+                i, Math.min(totalFrames - i, FRAME_CHUNK_READ_SIZE)));
+      } catch (IllegalStateException e) {
+        Log.d(ContentValues.TAG, "Video chunk read unsuccessful with: " + e.getMessage());
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /** Convenience method used to assign frame timestamps to individual frames based on video
+   * recording start time and the fps of video. */
+  private void setNextFrameTimeStamp() {
+    videoFrameTimestamp.add(recordStartTime + (framesProcessed * frameDuration));
+  }
+
+  public ArrayList<String> doOcr() {
+    AnalyseVideo ocrOnVideoTask = new AnalyseVideo();
+    try {
+      return ocrOnVideoTask.execute().get();
+    } catch (ExecutionException | InterruptedException e) {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  /** Iterates on video frames issuing an Ocr request. Requests video-frame timestamp assignment.
+   * Calls LagCalculator upon Ocr completion of all video frame.*/
+  private class AnalyseVideo extends AsyncTask<Void, Void, ArrayList<String>> {
+
+    @Override
+    protected ArrayList<String> doInBackground(Void ...values) {
+      ArrayList<String> resultsOCR = new ArrayList<>();
+      Collection<Task<Text>> ocrTasks = new ArrayList<>();
+      for (int i = 0; i < frameList.size(); i++) {
+        InputImage imageHolder = InputImage.fromBitmap(frameList.get(i), 0);
+        final int finalI = i;
+        ocrTasks.add(
+            recognizer
+                .process(imageHolder)
+                .addOnSuccessListener(
+                    visionText -> {
+                      resultsOCR.add(visionText.getText());
+                      Log.d(
+                          ContentValues.TAG,
+                          "Text detected at index:" + finalI + " " + visionText.getText());
+                      setNextFrameTimeStamp();
```

Comment:
This is a major concurrency problem.  You are calling this function which assumes it is called on frames in sequential order, however you are calling it from an async completion function which can be run for the frames out of order. In addition the accesses to member variables would need to be locked by mutex to prevent corruption.  You need to preserve the knowledge of which frame is doing this.

Also, is there any reason to do this here?  It could easily be done in the non-async part of the code right?

Reply:
Shifted to non-async part